File: objc/DKParser.m
//
//  DKParser.m
//  DavKit
//
//  Created by Aleksander Slater on 09/06/2012.
//  Copyright (c) 2012 Davincium. All rights reserved.
//

#import "DKParser.h"

#import "UIColor+DavKit.h"
#import "NSDate+DavKit.h"
#import "NSString+DavKit.h"

@implementation DKParser

+ (id)getObject:(NSDictionary*)d ofClass:(Class)classe forKey:(NSString*)key fallback:(id)fallBack
{
    NSString *obj=[d objectForKey:key];
    if (obj!=nil && [obj isKindOfClass:classe])
        return obj;
    return fallBack;
}

+ (id)getObject:(NSDictionary*)d ofClass:(Class)classe forKeys:(NSArray*)keys fallback:(id)fallBack
{
    for (NSString *key in keys)
    {
        NSString *obj=[d objectForKey:key];
        if (obj!=nil && [obj isKindOfClass:classe])
            return obj;
    }
    return fallBack;
}

+ (NSValue*)getValue:(NSDictionary*)d forKey:(NSString*)key fallback:(NSValue*)fallBack
{
    return [DKParser getObject:d ofClass:[NSValue class] forKey:key fallback:fallBack];
}

+ (NSNumber*)getNumber:(NSDictionary*)d forKey:(NSString*)key fallback:(NSNumber*)fallBack
{
    return [DKParser getObject:d ofClass:[NSNumber class] forKey:key fallback:fallBack];
}

+ (NSString*)getString:(NSDictionary*)d forKey:(NSString*)key fallback:(NSString*)fallBack
{
    return [DKParser getObject:d ofClass:[NSString class] forKey:key fallback:fallBack];
}

+ (NSArray*)getArray:(NSDictionary*)d forKey:(NSString*)key fallback:(NSArray*)fallBack
{
    return [DKParser getObject:d ofClass:[NSArray class] forKey:key fallback:fallBack];    
}

+ (NSDictionary*)getDictionary:(NSDictionary*)d forKey:(NSString*)key fallback:(NSDictionary*)fallBack
{
    return [DKParser getObject:d ofClass:[NSDictionary class] forKey:key fallback:fallBack];        
}

+ (int)getInt:(NSDictionary*)d forKey:(NSString*)key fallback:(int)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] intValue];
}

+ (NSInteger)getInteger:(NSDictionary*)d forKey:(NSString*)key fallback:(NSInteger)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] integerValue];
}

+ (long long)getLongLong:(NSDictionary*)d forKey:(NSString*)key fallback:(long long)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] longLongValue];
}

+ (unsigned long long)getUnsignedLongLong:(NSDictionary*)d forKey:(NSString*)key fallback:(unsigned long long)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] unsignedLongLongValue];
}

+ (BOOL)getBool:(NSDictionary*)d forKey:(NSString*)key fallback:(BOOL)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] boolValue];
}

+ (double)getDouble:(NSDictionary*)d forKey:(NSString*)key fallback:(double)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] doubleValue];
}

+ (CGFloat)getFloat:(NSDictionary*)d forKey:(NSString*)key fallback:(CGFloat)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] floatValue];
}

+ (CGFloat)getCGFloat:(nullable NSDictionary*)d forKey:(nonnull NSString*)key fallback:(CGFloat)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] floatValue];
}

+ (long)getLong:(NSDictionary*)d forKey:(NSString*)key fallback:(long)fallBack
{
    return [[DKParser getNumber:d forKey:key fallback:@(fallBack)] longValue];
}

+ (NSValue*)getValue:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(NSValue*)fallBack
{
    return [DKParser getObject:d ofClass:[NSValue class] forKeys:keys fallback:fallBack];
}

+ (NSNumber*)getNumber:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(NSNumber*)fallBack
{
    return [DKParser getObject:d ofClass:[NSNumber class] forKeys:keys fallback:fallBack];
}

+ (NSString*)getString:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(NSString*)fallBack
{
    return [DKParser getObject:d ofClass:[NSString class] forKeys:keys fallback:fallBack];
}

+ (NSArray*)getArray:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(NSArray*)fallBack
{
    return [DKParser getObject:d ofClass:[NSArray class] forKeys:keys fallback:fallBack];
}

+ (NSDictionary*)getDictionary:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(NSDictionary*)fallBack
{
    return [DKParser getObject:d ofClass:[NSDictionary class] forKeys:keys fallback:fallBack];
}

+ (NSInteger)getInteger:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(NSInteger)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] integerValue];
}

+ (int)getInt:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(int)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] intValue];
}

+ (long long)getLongLong:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(long long)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] longLongValue];
}

+ (BOOL)getBool:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(BOOL)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] boolValue];
}

+ (double)getDouble:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(double)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] doubleValue];
}

+ (CGFloat)getFloat:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(CGFloat)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] floatValue];
}

+ (long)getLong:(NSDictionary*)d forKeys:(NSArray*)keys fallback:(long)fallBack
{
    return [[DKParser getNumber:d forKeys:keys fallback:@(fallBack)] longValue];
}

+ (SEL)getSelector:(NSDictionary*)d forKey:(NSString*)key fallback:(SEL)fallBack
{
    NSValue *obj=[d objectForKey:key];
    if (obj!=nil && [obj isKindOfClass:[NSValue class]])
    {
        SEL sel = [obj pointerValue];
        return sel;
    }
    else if (obj!=nil && [obj isKindOfClass:[NSString class]])
    {
        NSString *str = (NSString*)obj;
        SEL sel = NSSelectorFromString(str);
        return sel;
    }
    return fallBack;
}

+ (BOOL)isDictionary:(id)d
{
    if (d!=nil && [d isKindOfClass:[NSDictionary class]])
        return YES;
    return NO;
}

+ (BOOL)isNumber:(id)number
{
    if (number!=nil && [number isKindOfClass:[NSNumber class]])
        return YES;
    return NO;
}

+ (BOOL)isString:(id)string
{
    if (string!=nil && [string isKindOfClass:[NSString class]] && [string length]>0)
        return YES;
    return NO;
}

+ (BOOL)isArray:(id)ar
{
    if (ar!=nil && [ar isKindOfClass:[NSArray class]])
        return YES;
    return NO;
}

+ (BOOL)isTrue:(id)o
{
    if ([o isKindOfClass:[NSString class]])
    {
        if ([o isEqualToString:@"1"])
            return YES;
        if ([[o lowercaseString] isEqualToString:@"true"])
            return YES;
    }
    if ([o isKindOfClass:[NSNumber class]])
    {
        NSInteger numb = [o integerValue];
        return numb;
    }
    return NO;
}

#pragma mark - Setters

+ (void)setObject:(id)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict fallback:(id)fallBack
{
    if (val==nil)
    {
        if (fallBack==nil)
        {
            [dict removeObjectForKey:key];
        }
        else
        {
            [dict setObject:fallBack forKey:key];
        }
    }
    else
    {
        [dict setObject:val forKey:key];
    }
}

+ (void)setNumber:(NSNumber*)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:val forKey:key inDict:dict fallback:nil];
}

+ (void)setString:(NSString*)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:val forKey:key inDict:dict fallback:nil];
}

+ (void)setDictionary:(NSDictionary*)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:val forKey:key inDict:dict fallback:nil];
}

+ (void)setArray:(NSArray*)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:val forKey:key inDict:dict fallback:nil];
}

+ (void)setInt:(int)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:[NSNumber numberWithInt:val] forKey:key inDict:dict fallback:nil];
}

+ (void)setInteger:(NSInteger)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:[NSNumber numberWithInteger:val] forKey:key inDict:dict fallback:nil];
}

+ (void)setLongLong:(long long)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:@(val) forKey:key inDict:dict fallback:nil];
}

+ (void)setUnsignedLongLong:(unsigned long long)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:@(val) forKey:key inDict:dict fallback:nil];
}

+ (void)setBool:(BOOL)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:[NSNumber numberWithBool:val] forKey:key inDict:dict fallback:nil];
}

+ (void)setSelector:(SEL)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:NSStringFromSelector(val) forKey:key inDict:dict fallback:nil];
}

#pragma mark - Containers

+ (NSInteger)addContentsOfDictInDict:(NSDictionary*)d withKey:(NSString*)key toDict:(NSMutableDictionary*)mdict
{
    NSDictionary *dict = [DKParser getDictionary:d forKey:key fallback:nil];
    if (dict!=nil)
    {
        [mdict addEntriesFromDictionary:dict];
    }
    return [dict count];
}

+ (NSInteger)addContentsOfArrayInDict:(NSDictionary*)d withKey:(NSString*)key toArray:(NSMutableArray*)mar
{
    NSArray *ar = [DKParser getArray:d forKey:key fallback:nil];
    if (ar!=nil)
    {
        [mar addObjectsFromArray:ar];
    }
    return [ar count];
}

#pragma mark - Special iOS and Mac

+ (void)setDate:(NSDate*)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    if (val==nil)
        [DKParser setObject:val forKey:key inDict:dict fallback:nil];
    else
        [DKParser setObject:[NSNumber numberWithDouble:[val timeIntervalSince1970]]
                     forKey:key
                     inDict:dict
                   fallback:nil];
}

+ (void)setRect:(CGRect)val forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [DKParser setObject:NSStringFromCGRect(val) forKey:key inDict:dict fallback:nil];
}

+ (NSDate*)getDate:(NSDictionary*)d forKey:(NSString*)key fallback:(NSDate*)fallBack
{
    NSNumber *n=[d objectForKey:key];
    if (n!=nil && [n isKindOfClass:[NSNumber class]])
        return [NSDate dateWithTimeIntervalSince1970:[n doubleValue]];
    if (n!=nil && [n isKindOfClass:[NSDate class]])
        return (NSDate*)n;
    return fallBack;
}

+ (CGRect)getRect:(NSDictionary*)d forKey:(NSString*)key fallback:(CGRect)fallBack
{
    NSString *n=[d objectForKey:key];
    if (n!=nil && [n isKindOfClass:[NSString class]])
        return CGRectFromString(n);
    return fallBack;
}

+ (void)setDateId:(NSDate*)date withId:(NSString*)Id forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    if (date==nil || Id==nil)
        [self setString:nil forKey:key inDict:dict];
    else
        [self setDateId:[[DKDateId alloc] initWithDate:date andId:Id] forKey:key inDict:dict];
}

+ (void)setDateId:(DKDateId*)dateId forKey:(NSString*)key inDict:(NSMutableDictionary*)dict
{
    [self setString:[dateId value] forKey:key inDict:dict];
}

+ (DKDateId*)getDateId:(NSDictionary*)d forKey:(NSString*)key fallback:(DKDateId*)fallBack
{
    NSString *s = [DKParser getString:d forKey:key fallback:nil];
    if (s!=nil)
    {
        DKDateId *dateId = [[DKDateId alloc] initWithString:s];
        if (dateId!=nil)
            return dateId;
    }
    return fallBack;
}

#pragma mark - Specials

+ (double)getDoubleFromString:(NSDictionary*)d forKey:(NSString*)key startsWith:(NSString*)start endsWith:(NSString*)end fallback:(double)fallBack
{
    NSString *s=[DKParser getString:d forKey:key fallback:nil];
    if (s!=nil)
    {
        if (start!=nil && end!=nil)
        {
            if ([s hasPrefix:start] && [s hasSuffix:end])
            {
                NSRange rstart = [s rangeOfString:start];
                NSRange rend = [s rangeOfString:end];
                NSRange res; res.location = rstart.location+rstart.length; res.length=rend.location-res.location;
                NSString *numString = [s substringWithRange:res];
                return [numString doubleValue];
            }
        }
    }
    return fallBack;
}

+ (double)getDoubleFromDotNetDateString:(NSDictionary*)d forKey:(NSString*)key fallback:(double)fallBack
{
    return [DKParser getDoubleFromString:d  forKey:key startsWith:@"/Date(" endsWith:@")/" fallback:fallBack];
}

+ (unsigned int)intFromHexString:(NSString *)hexStr
{
    unsigned int hexInt = 0;
    
    // Create scanner
    NSScanner *scanner = [NSScanner scannerWithString:hexStr];
    
    // Tell scanner to skip the # character
    [scanner setCharactersToBeSkipped:[NSCharacterSet characterSetWithCharactersInString:@"#"]];
    
    // Scan hex value
    [scanner scanHexInt:&hexInt];
    
    return hexInt;
}

+ (NSDictionary*)getDictionaryFromUrl:(NSURL*)url
{
    NSMutableDictionary *params = [[NSMutableDictionary alloc] init];
    NSString *subString = [url query];
    NSArray *pairs = [subString componentsSeparatedByString:@"&"];
    for (NSString *pair in pairs)
    {
        NSArray *kv = [pair componentsSeparatedByString:@"="];
        NSString *val =[[kv objectAtIndex:1] stringByReplacingPercentEscapes];
        [params setObject:val forKey:[kv objectAtIndex:0]];
    }
    NSDictionary *returnDict = [NSDictionary dictionaryWithDictionary:params];
    return returnDict;
}

#pragma mark - JSON

+ (id)fromJSONString:(NSString*)jsonString
{
    NSData *data = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
    return [DKParser fromJSONData:data];
}

+ (id)fromJSONData:(NSData*)data
{
    if (data!=nil)
    {
        id json = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        return json;
    }
    return nil;
}

+ (NSString*)toJSONString:(id)object
{
    NSError *error = nil;
    
    NSData *data = [NSJSONSerialization dataWithJSONObject:object options:0 error:&error];
    if (data!=nil)
    {
        NSString *str = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
        return str;
    }
    return nil;
}

+ (NSDictionary*)fromJSONToDictionary:(NSString*)jsonString
{
    return [DKParser fromJSONString:jsonString];
}

+ (NSArray*)fromJSONToArray:(NSString*)jsonString
{
    return [DKParser fromJSONString:jsonString];
}

+ (NSString*)fromDictionaryToJSON:(NSDictionary*)d
{
    return [DKParser toJSONString:d];
}

+ (NSString*)fromArrayToJSON:(NSArray*)d
{
    return [DKParser toJSONString:d];
}

+ (NSDictionary*)jsonSafeDictionary:(NSDictionary*)d
{
    if (d==nil) return nil;
    NSMutableDictionary *md = [[NSMutableDictionary alloc] init];
    for (id key in d.allKeys)
    {
        id val = [d objectForKey:key];
        if ([val isKindOfClass:[NSDate class]])
        {
            [md setObject:@([((NSDate*)val) timeIntervalSince1970]) forKey:key];
        }
        else if ([val isKindOfClass:[UIColor class]])
        {
            [md setObject:[val hexRGBAString] forKey:key];
        }
        else if ([val isKindOfClass:NSClassFromString(@"DBList")])
        {
            [md setObject:[val values] forKey:key];
        }
        else
        {
            [md setObject:val forKey:key];
        }
    }
    return md;
}

+ (NSDictionary*)jsonSafeDictionaryISO:(NSDictionary*)d
{
    if (d==nil) return nil;
    NSMutableDictionary *md = [[NSMutableDictionary alloc] init];
    for (id key in d.allKeys)
    {
        id val = [d objectForKey:key];
        if ([val isKindOfClass:[NSDate class]])
        {
            [md setObject:[val ISO8601String] forKey:key];
        }
        else if ([val isKindOfClass:[UIColor class]])
        {
            [md setObject:[val hexRGBAString] forKey:key];
        }
        else if ([val isKindOfClass:NSClassFromString(@"DBList")])
        {
            [md setObject:[val values] forKey:key];
        }
        else
        {
            [md setObject:val forKey:key];
        }
    }
    return md;
}

+ (NSDictionary*)convertDictValues:(NSDictionary*)d usingFormatters:(NSDictionary*)formatters
{
    if (formatters==nil || d==nil)
        return d;
    
    NSMutableDictionary *md = [[NSMutableDictionary alloc] initWithDictionary:d];
    for (id fkey in formatters.allKeys)
    {
        Class c = [d objectForKey:fkey];
        if ([NSDate class] == c)
        {
            NSDate *date = [DKParser getDate:md forKey:fkey fallback:nil];
            if (date!=nil)
            {
                [md setObject:date forKey:fkey];
            }
        }
    }
    return md;
}

@end

@implementation DKDateId

- (instancetype)initWithString:(NSString*)s
{
    self = [super init];
    if (self)
    {
        self.value = s;
    }
    
    if (self.Id!=nil && self.date!=nil)
        return self;
    return nil;
}

- (instancetype)initWithDate:(NSDate*)aDate andId:(NSString*)anId
{
    self = [super init];
    if (self)
    {
        [self setDate:aDate andId:anId];
    }
    return self;
}

- (void)setDate:(NSDate*)aDate andId:(NSString*)anId
{
    self.value = [NSString stringWithFormat:@"%f|%@",[aDate timeIntervalSince1970],anId];
}

- (NSArray*)components
{
    return [self.value componentsSeparatedByString:@"|"];
}

- (NSString*)Id
{
    NSArray *c = [self components];
    if (c.count==2)
    {
        return [c objectAtIndex:1];
    }
    return nil;
}

- (NSDate*)date
{
    NSArray *c = [self components];
    if (c.count==2)
    {
        NSString *c1 = [c firstObject];
        NSTimeInterval interval = [c1 doubleValue];
        return [NSDate dateWithTimeIntervalSince1970:interval];
    }
    return nil;
}

@end

File: swift/DKParser.swift

import Foundation
import UIKit

protocol IDKList : NSObjectProtocol
{
    func values() -> NSArray?
}

public class DKParser
{
    public class func getObject(d:NSDictionary?,ofClass classe:AnyClass,forKey key:String,fallback:AnyObject?) -> AnyObject?
    {
        if let od = d
        {
            if let obj = od.objectForKey(key)
            {
                if obj.isKindOfClass(classe) { return obj }
            }
        }
        return fallback
    }
    
    public class func getObject(d:NSDictionary?,ofClass classe:AnyClass,forKeys keys:Array<String>,fallback:AnyObject?) -> AnyObject?
    {
        for key in keys
        {
            if let od = d
            {
                if let obj = od.objectForKey(key)
                {
                    if obj.isKindOfClass(classe) { return obj }
                }
            }
        }
        return fallback
    }
    
    public class func getSafeNumber(d:NSDictionary?,forKey key:String, fallback:NSNumber) -> NSNumber
    {
        if let ret = getNumber(d, forKey: key, fallback: fallback)
        {
            return ret
        }
        return fallback
    }
    
    public class func getSafeNumber(d:NSDictionary?,forKeys keys:Array<String>,fallback:NSNumber) -> NSNumber
    {
        if let ret = getNumber(d, forKeys: keys, fallback: fallback)
        {
            return ret
        }
        return fallback
    }
    
    public class func getValue(d:NSDictionary?,forKey key:String, fallback:NSValue?) -> NSValue?
    {
        return getObject(d,ofClass:NSValue.self,forKey:key,fallback:fallback) as? NSValue
    }
    
    public class func getArray(d:NSDictionary?,forKey key:String, fallback:NSArray?) -> NSArray?
    {
        return getObject(d,ofClass:NSArray.self,forKey:key,fallback:fallback) as? NSArray
    }
    
    public class func getDictionary(d:NSDictionary?,forKey key:String, fallback:NSDictionary?) -> NSDictionary?
    {
        return getObject(d,ofClass:NSDictionary.self,forKey:key,fallback:fallback) as? NSDictionary
    }
    
    public class func getString(d:NSDictionary?,forKey key:String, fallback:String?) -> String?
    {
        return getObject(d,ofClass:object_getClass(""),forKey:key,fallback:fallback) as? String
    }
    
    public class func getNumber(d:NSDictionary?,forKey key:String, fallback:NSNumber?) -> NSNumber?
    {
        return getObject(d,ofClass:NSNumber.self,forKey:key,fallback:fallback) as? NSNumber
    }
    
    public class func getInteger(d:NSDictionary?,forKey key:String, fallback:Int) -> Int
    {
        return getSafeNumber(d, forKey: key, fallback:fallback).integerValue
    }
    
    public class func getBool(d:NSDictionary?,forKey key:String, fallback:Bool) -> Bool
    {
        return getSafeNumber(d, forKey: key, fallback:fallback).boolValue
    }
    
    public class func getFloat(d:NSDictionary?,forKey key:String, fallback:Float) -> Float
    {
        return getSafeNumber(d, forKey: key, fallback:fallback).floatValue
    }
    
    public class func getDouble(d:NSDictionary?,forKey key:String, fallback:Double) -> Double
    {
        return getSafeNumber(d, forKey: key, fallback:fallback).doubleValue
    }
    
    public class func getInt(d:NSDictionary?,forKey key:String, fallback:Int32) -> Int32
    {
        return getSafeNumber(d, forKey: key, fallback:NSNumber(int:fallback)).intValue
    }
    
    public class func getLongLong(d:NSDictionary?,forKey key:String, fallback:Int64) -> Int64
    {
        return getSafeNumber(d, forKey: key, fallback:NSNumber(longLong:fallback)).longLongValue
    }
    
    public class func getValue(d:NSDictionary?,forKeys keys:Array<String>, fallback:NSValue?) -> NSValue?
    {
        return getObject(d,ofClass:NSValue.self,forKeys:keys,fallback:fallback) as? NSValue
    }
    
    public class func getArray(d:NSDictionary?,forKeys keys:Array<String>, fallback:NSArray?) -> NSArray?
    {
        return getObject(d,ofClass:NSArray.self,forKeys:keys,fallback:fallback) as? NSArray
    }
    
    public class func getDictionary(d:NSDictionary?,forKeys keys:Array<String>, fallback:NSDictionary?) -> NSDictionary?
    {
        return getObject(d,ofClass:NSDictionary.self,forKeys:keys,fallback:fallback) as? NSDictionary
    }
    
    public class func getString(d:NSDictionary?,forKeys keys:Array<String>, fallback:String?) -> String?
    {
        return getObject(d,ofClass:object_getClass(""),forKeys:keys,fallback:fallback) as? String
    }
    
    public class func getNumber(d:NSDictionary?,forKeys keys:Array<String>, fallback:NSNumber?) -> NSNumber?
    {
        return getObject(d,ofClass:NSNumber.self,forKeys:keys,fallback:fallback) as? NSNumber
    }
    
    public class func getInteger(d:NSDictionary?,forKeys keys:Array<String>, fallback:Int) -> Int
    {
        return getSafeNumber(d, forKeys:keys, fallback:fallback).integerValue
    }
    
    public class func getBool(d:NSDictionary?,forKeys keys:Array<String>, fallback:Bool) -> Bool
    {
        return getSafeNumber(d, forKeys: keys, fallback:fallback).boolValue
    }
    
    public class func getFloat(d:NSDictionary?,forKeys keys:Array<String>, fallback:Float) -> Float
    {
        return getSafeNumber(d, forKeys: keys, fallback:fallback).floatValue
    }
    
    public class func getDouble(d:NSDictionary?,forKeys keys:Array<String>, fallback:Double) -> Double
    {
        return getSafeNumber(d, forKeys: keys, fallback:fallback).doubleValue
    }
    
    public class func getInt(d:NSDictionary?,forKeys keys:Array<String>, fallback:Int32) -> Int32
    {
        return getSafeNumber(d, forKeys: keys, fallback:NSNumber(int:fallback)).intValue
    }
    
    public class func getLongLong(d:NSDictionary?,forKeys keys:Array<String>, fallback:Int64) -> Int64
    {
        return getSafeNumber(d, forKeys: keys, fallback:NSNumber(longLong:fallback)).longLongValue
    }
    
    public class func getDate(d:NSDictionary?,forKey key:String,fallback:NSDate?) -> NSDate?
    {
        if let n = d?.objectForKey(key) as? NSNumber
        {
            return NSDate(timeIntervalSince1970:n.doubleValue)
        }
        if let date = d?.objectForKey(key) as? NSDate
        {
            return date
        }
        return fallback
    }
    
    public class func getDate(d:NSDictionary?,forKeys keys:Array<String>,fallback:NSDate?) -> NSDate?
    {
        for key in keys
        {
            if let od = d
            {
                if let n = od.objectForKey(key) as? NSNumber
                {
                    return NSDate(timeIntervalSince1970:n.doubleValue)
                }
                if let date = od.objectForKey(key) as? NSDate
                {
                    return date
                }
            }
        }
        return fallback
    }
    
    public class func getList(d:NSDictionary?,forKey key:String, fallback:NSArray?) -> NSArray?
    {
        if let classe = NSClassFromString("DBList")
        {
            if let obj = getObject(d,ofClass:classe,forKey:key,fallback:nil) as? IDKList
            {
                return obj.values()
            }
        }
        return getArray(d,forKey:key,fallback:fallback)
    }
    
    // MARK: Setters
    
    public class func setObject(val:AnyObject?,forKey key:String,inDict dict:NSMutableDictionary?,replacement:AnyObject?)
    {
        if let v = val
        {
            dict?.setObject(v,forKey:key)
        }
        else
        {
            if let fb = replacement
            {
                dict?.setObject(fb,forKey:key)
            }
            else
            {
                dict?.removeObjectForKey(key)
            }
        }
    }
    
    public class func setNumber(val:NSNumber?,forKey key:String,inDict dict:NSMutableDictionary?,replacement:NSNumber?=nil)
    {
        setObject(val, forKey:key, inDict:dict, replacement:replacement)
    }
    
    public class func setString(val:String?,forKey key:String,inDict dict:NSMutableDictionary?,replacement:String?=nil)
    {
        setObject(val, forKey:key, inDict:dict, replacement:replacement)
    }
    
    public class func setDictionary(val:NSDictionary?,forKey key:String,inDict dict:NSMutableDictionary?,replacement:NSDictionary?=nil)
    {
        setObject(val, forKey:key, inDict:dict, replacement:replacement)
    }
    
    public class func setArray(val:NSArray?,forKey key:String,inDict dict:NSMutableDictionary?,replacement:NSArray?=nil)
    {
        setObject(val, forKey:key, inDict:dict, replacement:replacement)
    }
    
    public class func setInt(val:Int32,forKey key:String,inDict dict:NSMutableDictionary?)
    {
        setObject(NSNumber(int:val), forKey:key, inDict:dict, replacement:nil)
    }
    
    public class func setInteger(val:Int,forKey key:String,inDict dict:NSMutableDictionary?)
    {
        setObject(NSNumber(integer:val), forKey:key, inDict:dict, replacement:nil)
    }
    
    public class func setBool(val:Bool,forKey key:String,inDict dict:NSMutableDictionary?)
    {
        setObject(NSNumber(bool:val), forKey:key, inDict:dict, replacement:nil)
    }
    
    public class func setDouble(val:Double,forKey key:String,inDict dict:NSMutableDictionary?)
    {
        setObject(NSNumber(double:val), forKey:key, inDict:dict, replacement:nil)
    }
    
    public class func setFloat(val:Float,forKey key:String,inDict dict:NSMutableDictionary?)
    {
        setObject(NSNumber(float:val), forKey:key, inDict:dict, replacement:nil)
    }
    
    public class func setSelector(val:Selector?,forKey key:String,inDict dict:NSMutableDictionary?)
    {
        if let s = val
        {
            setObject(NSStringFromSelector(s), forKey:key, inDict:dict, replacement:nil)
        }
        else
        {
            setObject(nil, forKey:key, inDict:dict, replacement:nil)
        }
    }
    
    public class func setDate(val:NSDate?,forKey key:String,inDict dict:NSMutableDictionary?,replacement:NSDate?=nil)
    {
        if let v = val
        {
            setObject(NSNumber(double:v.timeIntervalSince1970), forKey:key, inDict:dict, replacement:replacement)
        }
        else
        {
            setObject(val, forKey:key, inDict:dict, replacement:replacement)
        }
    }
}
